contracts/ProofOfHumanity.sol

```diff
@@ -469,15 +469,52 @@ contract ProofOfHumanity is IArbitrable, IEvidence {
     /** @dev Change submission's state from Vouching to PendingRegistration if all conditions are met.
      *  @param _submissionID The address of the submission which status to change.
      *  @param _vouches Array of users which vouches to count.
+     *  @param _signatures Array of signatures (optional)
+     *  @param _timeouts Array of timeouts for each signature (optional)
+     *  The signature format is "ProofOfHumanity<20:submissionID><32:signatureTimeout>"
      */
-    function changeStateToPending(address _submissionID, address[] calldata _vouches) external {
+    function changeStateToPending(address _submissionID, address[] calldata _vouches, bytes[] calldata _signatures, uint[] calldata _timeouts) external {
         Submission storage submission = submissions[_submissionID];
         require(submission.status == Status.Vouching, "Wrong status");
         Request storage request = submission.requests[submission.requests.length - 1];
         Challenge storage challenge = request.challenges[0];
         Round storage round = challenge.rounds[0];
         require(round.hasPaid[uint(Party.Requester)], "Requester is not funded");
 
+        bytes memory PREFIX = "\x19Ethereum Signed Message:\n67ProofOfHumanity";
+        for (uint i = 0; i < _signatures.length && request.vouches.length < requiredNumberOfVouches; i++) {
+            address voucherAddress;
+            /* solium-disable indentation */
+            {
+                // Get message hash
+                bytes32 hash = keccak256(abi.encodePacked(PREFIX, _submissionID, _timeouts[i]));
+
+                // Decode the signature
+                bytes memory signature = _signatures[i];
+                bytes32 r;
+                bytes32 s;
+                uint8 v;
+                assembly {
+                    r := mload(add(signature, 0x20))
+                    s := mload(add(signature, 0x40))
+                    v := byte(0, mload(add(signature, 0x60)))
+                }
+                if (v < 27) v += 27;
+                require(v == 27 || v == 28, "Invalid signature");
+
+                // Recover the signer's address
+                voucherAddress = ecrecover(hash, v, r, s);
+            }
+            /* solium-enable indentation */
+
+            Submission storage voucher = submissions[voucherAddress];
+            if (!voucher.hasVouched && voucher.registered && now - voucher.submissionTime <= submissionDuration &&
+            now < _timeouts[i] && _submissionID != voucherAddress) {
+                request.vouches.push(voucherAddress);
+                voucher.hasVouched = true;
+            }
+        }
+
         for (uint i = 0; i<_vouches.length && request.vouches.length<requiredNumberOfVouches; i++) {
```

Comment:
If `requiredNumberOfVouches > _signatures.length`, `request.vouches` will have unnecessary extra vouches.

Reply:
There is request.vouches.length < requiredNumberOfVouches in both loops, how could this happen?

Reply:
Ahh, missed that.